File: day01/Day01.hs
#!/usr/bin/env runghc

module Day01 where

import Data.List

main :: IO ()
main = do
  contents <- readFile "day01/input"
  print $ getMaxSum contents
  print $ get3MaxSum contents

getMaxSum :: String -> Integer
getMaxSum = maximum . map sum . parse

get3MaxSum :: String -> Integer
get3MaxSum = (*) (-1) . sum . take 3 . sort . map ((*) (-1) . sum) . parse

parse :: String -> [[Integer]]
parse = map (map read) . split (== "") . lines

split :: (a -> Bool) -> [a] -> [[a]]
split _ [] = []
split check l@(x:xs) =
  if check x then split check xs else
    let heads = takeWhile (not . check) l in
      let tails = dropWhile (not . check) l in
        heads : split check tails

File: day01/day01.py
#!/usr/bin/env python

def main():
    with open("day01/input", "r") as fin:
        contents = fin.read()
    print(get_max_sum(contents))
    print(get3_max_sum(contents))


def get_max_sum(data):
    return max(map(sum, parse(data)))

def get3_max_sum(data):
    vals = map(sum, parse(data))
    first = second = third = 0
    for v in vals:
        if v <= third:
            continue
        third = v
        if third <= second:
            continue
        second, third = third, second
        if second <= first:
            continue
        first, second = second, first
    return first + second + third


def parse(data):
    lines = data.split("\n")
    groups = []
    current = []
    for line in lines:
        if line == "":
            groups.append(current)
            current = []
        else:
            current.append(int(line))
    return groups


if __name__ == "__main__":
    main()
